### PR TITLE
Defensively check the dereferenced value is not a resource

### PR DESCRIPTION
**Pull request by turbolent**
______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

**Review by dsainati1 — COMMENTED** [review]
Can we add an interpreter test here to exercise this (presumably by ignoring the checker error)?

File: runtime/interpreter/value.go [hunk 1]
```diff
@@ -19757,8 +19757,14 @@ func DereferenceValue(
 	locationRange LocationRange,
 	referenceValue ReferenceValue,
 ) Value {
-	referencedValue := referenceValue.ReferencedValue(inter, locationRange, true)
-	return (*referencedValue).Transfer(
+	referencedValue := *referenceValue.ReferencedValue(inter, locationRange, true)
+
+	// Defensive check: ensure that the referenced value is not a resource
+	if referencedValue.IsResourceKinded(inter) {
+		panic(errors.NewUnreachableError())
```
Comment (SupunS) [inline runtime/interpreter/value.go:19764]: Maybe throw a dedicated error

Comment (turbolent) [reply runtime/interpreter/value.go]: Good point, improved in 34e9e43d1e6b3ad0a4bf34b26a621cdb097360a5

**Comment by turbolent** [comment]
@dsainati1 Added tests in 26d4494464957d5ae467b1b80b5c7af18654b23b

**Review by SupunS — APPROVED** [review]

**Pull request merged**